Multi auth methods

arthurpassos opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow a user to have multiple authentication methods instead of only one. Allow authentication methods to be reset to most recently added method. If you want to run instances on 24.8 and one on 24.9 for some time, it's better to set `max_authentication_methods_per_user = 1` for that period to avoid potential errors.

`CREATE USER user1 IDENTIFIED WITH plaintext_password BY '1', plaintext_password BY '2', sha256_password BY '3'` -- user can authenticate with `1`, `2` or `3`. 

Same logic applies to `ALTER` statements.

`ALTER USER user1 RESET AUTHENTICATION METHODS TO NEW` -- clears all authentication methods and keeps only the most recent

`ALTER USER user1 IDENTIFIED WITH plaintext_password by '1'` -- should remain working as it works nowadays, it'll clear all authentication methods and keep only the on found in the query.

Closes https://github.com/ClickHouse/ClickHouse/issues/65136

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_normal_builds--> Allow: Normal Builds
- [ ] <!---ci_set_special_builds--> Allow: Special Builds
- [ ] <!---ci_set_non_required--> Allow: All NOT Required Checks
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

arthurpassos committed:
testing the waters

arthurpassos committed:
throw exception upon auth

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #65277 from arthurpassos/ClickHouse:multi_auth_methods -->
*This is an automated comment for commit b22776d3a81784daf23ac6e04871912249695030 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65277/b22776d3a81784daf23ac6e04871912249695030/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

arthurpassos committed:
style fix

arthurpassos committed:
append default constructed auth method upon alter without auth data

arthurpassos committed:
fix no pwd authentication

arthurpassos committed:
fix test build

Avogar commented:
Test `03036_dynamic_read_subcolumns` was failing in master, but it's fixed. Please, update the branch with recent changes in master

alexey-milovidov commented:
This is from another PR, but I just leave it here.
A test for multiple PTR records has failed: https://s3.amazonaws.com/clickhouse-test-reports/60197/11d54f4809a8b58773f13664e6b842cb6c7dce48/integration_tests__aarch64__[2_6].html

arthurpassos committed:
ptal

arthurpassos committed:
make progress, seems functional

arthurpassos commented:
> This is from another PR, but I just leave it here. A test for multiple PTR records has failed: [https://s3.amazonaws.com/clickhouse-test-reports/60197/11d54f4809a8b58773f13664e6b842cb6c7dce48/integration_tests__aarch64__[2_6].html](https://s3.amazonaws.com/clickhouse-test-reports/60197/11d54f4809a8b58773f13664e6b842cb6c7dce48/integration_tests__aarch64__%5B2_6%5D.html)

@alexey-milovidov Will look into it soon

arthurpassos committed:
simplify syntax

arthurpassos commented:
@alexey-milovidov do you think we could divide this work into two PRs? First one to support multiple authentication methods and second one to add expiration time?

alexey-milovidov commented:
Totally ok, but we already have support for expiration time (for about half a year).

arthurpassos committed:
fix wrong condition

arthurpassos committed:
fix astcreateuserquery clone

arthurpassos committed:
add some ut

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
initial tests

arthurpassos committed:
use plaintext_password instead of sha256 because of cicd

arthurpassos committed:
do not gen at runtime, use pre-built ones

arthurpassos committed:
dont test ssh at all, it wont work if openssl is not built

arthurpassos committed:
remove comments

arthurpassos commented:
Removed stateless ssh tests because it depends on openssl being built

arthurpassos committed:
Revert "dont test ssh at all, it wont work if openssl is not built"

This reverts commit 4d5676f4551310380d7cad6e012bb155a2843524.

arthurpassos committed:
no-fast test

arthurpassos committed:
add no parallel

arthurpassos committed:
serialize no_password

arthurpassos commented:
@alexey-milovidov https://github.com/ClickHouse/ClickHouse/pull/65277/commits/57afc5f0353ce7b7e6251940f2d17c40b1480855 serializes `AuthenticationType::NO_PASSWORD`.

Without this commit, `NO_PASSWORD` is omitted during serialization and it is inferred during startup that it uses no_password if there is no authentication method serialized. This is a problem if we want to support the below:

`CREATE USER user1;`
`ALTER USER user1 ADD IDENTIFIED WITH plaintext_password BY '1'` -- add new authentication method

`clickhouse-client --user user1` -- succeeds, `NO_PASSWORD` is still active

Restart ClickHouse. During startup, it finds an authentication method, so it assumes that is the only authentication method.

`clickhouse-client --user user1` -- **fails to authenticate with previously registered no_password**

We could also forbid `NO_PASSWORD` to co-exist with any other authentication method, then this commit is not necessary I believe.

alexey-milovidov commented:
Yes, serializing NO_PASSWORD is a good change.

arthurpassos committed:
update some tests

arthurpassos committed:
update some more tests

arthurpassos committed:
update test_disk_access_storage

arthurpassos committed:
add no-parallel to grant_and_Revoke.sql

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
black

arthurpassos committed:
fix wrong expected error code & add test

arthurpassos committed:
update some other tests

arthurpassos committed:
black

arthurpassos committed:
update yet another test

arthurpassos committed:
initial fix for session_log, shall be refactored

arthurpassos committed:
black

arthurpassos committed:
update some other tests

arthurpassos committed:
try to fix black

arthurpassos committed:
try to fix black

arthurpassos committed:
try to fix black

arthurpassos committed:
small refactor

arthurpassos committed:
minor

arthurpassos committed:
remove definition of logical_error from auth.cpp

arthurpassos committed:
remove optional from auth_result

arthurpassos committed:
remmovev optional from auth data in session object

arthurpassos commented on src/Interpreters/Session.h:
```diff
@@ -113,6 +113,7 @@ class Session
 
     mutable UserPtr user;
     std::optional<UUID> user_id;
+    AuthenticationData user_authenticated_with;
```
I am not quite familiar with the `Session` object lifetime, but this field is filled upon authentication and is only used in `recordLoginSuccess`, so I assume it is safe to use it as is.

arthurpassos commented on src/Parsers/Access/ASTCreateUserQuery.cpp:
```diff
@@ -18,9 +18,17 @@ namespace
+        auth_data[0]->format(settings);
+
+        auto settings_with_additional_authentication_method = settings;
+        settings_with_additional_authentication_method.additional_authentication_method = true;
```
I don't like this approach very much, ideas on how to improve it? One option would be to pollute the `ASTAuthenticationData` and `AuthenticationData` with a field like `bool is_add_identified`. I don't like this approach very much.

pufit replied:
The easiest approach I can see is to do something like
```cpp
for (size_t i = 1; i < authentication_methods.size(); ++i)
{
    settings.ostr << (settings.hilite ? IAST::hilite_keyword : "") << " ADD" << (settings.hilite ? IAST::hilite_none : "");
    authentication_methods[i]->format(settings);
}
```

arthurpassos committed:
rename somem stuff

arthurpassos committed:
trigger ci

arthurpassos committed:
fix integ test

arthurpassos committed:
add missing reference qualifier

arthurpassos commented:
Stress test failure - https://github.com/ClickHouse/ClickHouse/issues/51339

arthurpassos committed:
add some docs

arthurpassos committed:
add a few comments

arthurpassos commented:
**[UNRELATED TO THIS PR]**

> This is from another PR, but I just leave it here. A test for multiple PTR records has failed: [https://s3.amazonaws.com/clickhouse-test-reports/60197/11d54f4809a8b58773f13664e6b842cb6c7dce48/integration_tests__aarch64__[2_6].html](https://s3.amazonaws.com/clickhouse-test-reports/60197/11d54f4809a8b58773f13664e6b842cb6c7dce48/integration_tests__aarch64__%5B2_6%5D.html)

I briefly looked into it and here's what I gathered:

It is attempting to perform 5k `SELECT 1` requests (100 threads, 50 iterations each) that force DNS resolution to happen through `host_regexp`. Looking at the ClickHouse logs, I see 4855 `Authenticating with user test_dns` :

```
arthur@arthur:~/tmp$ grep -c "Authenticating user 'test_dns'" log 
4855
```

I also see 4855 successfull queries on CoreDNS logs:

```
arthur@arthur:~/tmp$ grep -c "PTR IN 3.15.16.172.in-addr.arpa. udp 42 false 512\" NOERROR" docker.log
4855
```

There is no `Authentication failed` in the logs:

```
arthur@arthur:~/tmp$ grep -c "Authentication failed" log 
0
```

There is not much material to work with, but I would guess CH couldn't handle this many requests concurrently. I can either remove the test completely or reduce the number of threads from 100 to a 50. Wdyt?

@alexey-milovidov

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

pufit commented on docs/en/sql-reference/statements/alter/user.md:
````diff
@@ -62,3 +62,20 @@ Allows the user with `john` account to grant his privileges to the user with `ja
+
+Reset authentication methods and keep the most recent added one
+``` sql
+ALTER USER user RESET AUTHENTICATION METHODS TO NEW
````
This looks counterintuitive. In what scenario can this command be helpful?

arthurpassos replied:
This comes from both Milovidov suggestion in https://github.com/ClickHouse/ClickHouse/issues/65136#issuecomment-2162096010 and https://www.percona.com/blog/using-mysql-8-dual-passwords/

pufit commented on src/Access/IAccessStorage.cpp:
```diff
@@ -525,15 +525,24 @@ std::optional<AuthResult> IAccessStorage::authenticateImpl(
+                auto auth_type = auth_method.getType();
+                if (((auth_type == AuthenticationType::NO_PASSWORD) && !allow_no_password) ||
+                    ((auth_type == AuthenticationType::PLAINTEXT_PASSWORD) && !allow_plaintext_password))
+                    throwAuthenticationTypeNotAllowed(auth_type);
```
Minor note: if a user has a plaintext password and any other valid auth method, this will result in an error without trying a valid one.

arthurpassos replied:
Yeah. I could also try to deny it only if the auth method was plaintext_password. What do you prefer?

arthurpassos replied:
Ok, it's not so trivial to do it. `BasicCredentials` does not have a field to indicate wheather the provided credentials are plaintext.

pufit commented on src/Parsers/Access/ASTCreateUserQuery.cpp:
```diff
@@ -19,9 +19,17 @@ namespace
+        auto settings_with_additional_authentication_method = settings;
+        settings_with_additional_authentication_method.additional_authentication_method = true;
+
+        for (auto i = 1u; i < authentication_methods.size(); i++)
```
```suggestion
        for (size_t i = 1; i < authentication_methods.size(); ++i)
```

arthurpassos committed:
simplify formatting of astauth

alexey-milovidov commented:
It is ok to remove that test (in a separate PR).

arthurpassos commented:
> It is ok to remove that test (in a separate PR).

@alexey-milovidov https://github.com/ClickHouse/ClickHouse/pull/66233

arthurpassos commented:
@pufit @alexey-milovidov What is your take on `NO_PASSWORD` authentication methods co-existing with other authentication methods? Does it make sense to allow a user to be logged in without a password if other authentication methods were added?

```
CREATE USER user1;
ALTER USER user1 ADD IDENTIFIED WITH plaintext_password by '123';
clickhouse-client --user user1
```

This also becomes a problem if we have `allow_no_password=0`. There is no way to look at the `BasicCredentials` object and distinguish it between NO_PASSWORD and PLAINTEXT_PASSWORD. All that object offers is a `getPassword` method. The same applies to `allow_plaintext_password`. With this PR, upon authentication, if a user has `allow_no_password=0` and one of the authentication methods is of type `NO_PASSWORD`, it will not authenticate even if the given credentials is of `PLAINTEXT_PASSWORD`, for instance.

One way to solve this would be to drop `NO_PASSWORD` authentication method if a different method is added. This would make sure that `NO_PASSWORD` does not co-exist with other authentication methods and is always the only authentication method, making it safe to throw exception and deny authentication. But then another problem occurs: what if the user adds `NO_PASSWORD` to a user that has other authentication methods? Do we drop the existing methods in favor of `NO_PASSWORD`? I assume yes.

arthurpassos committed:
allow other auth methods to be used if no_password is setup but not allowed

arthurpassos commented:
> @pufit @alexey-milovidov What is your take on `NO_PASSWORD` authentication method co-existing with other authentication methods? Does it make sense to allow a user to be logged in without a password if other authentication methods were added?
> 
> ```
> CREATE USER user1;
> ALTER USER user1 ADD IDENTIFIED WITH plaintext_password by '123';
> clickhouse-client --user user1
> ```
> 
> This also becomes a problem if we have `allow_no_password=0`. There is no way to look at the `BasicCredentials` object and distinguish it between NO_PASSWORD and PLAINTEXT_PASSWORD. All that object offers is a `getPassword` method. The same applies to `allow_plaintext_password`. With this PR, upon authentication, if a user has `allow_no_password=0` and one of the authentication methods is of type `NO_PASSWORD`, it will not authenticate even if the given credentials is of `PLAINTEXT_PASSWORD`, for instance.
> 
> One way to solve this would be to drop `NO_PASSWORD` authentication method if a different method is added. This would make sure that `NO_PASSWORD` does not co-exist with other authentication methods and is always the only authentication method, making it safe to throw exception and deny authentication. But then another problem occurs: what if the user adds `NO_PASSWORD` to a user that has other authentication methods? Do we drop the existing methods in favor of `NO_PASSWORD`? I assume yes.

In https://github.com/ClickHouse/ClickHouse/pull/65277/commits/cd0145113f751228e56500f6d2907433929e9622 I changed the exception logic upon authentication from: deny access if user has `no_password` registered and `allow_no_password=0` even if credentials are valid to: simply skip such authentication methods. The downside is that the error is generic, but I could also a warning: some auth methods were skipped.

pufit commented:
> One way to solve this would be to drop NO_PASSWORD authentication method if a different method is added. This would make sure that NO_PASSWORD does not co-exist with other authentication methods and is always the only authentication method, making it safe to throw exception and deny authentication. 

I personally believe that this should be the best approach

> But then another problem occurs: what if the user adds NO_PASSWORD to a user that has other authentication methods? Do we drop the existing methods in favor of NO_PASSWORD? I assume yes.

We should not allow adding the NO_PASSWORD method to existing ones, so an error should probably be enough. If a user wants to set NO_PASSWORD, they should override auth methods explicitly.

arthurpassos committed:
remove unused extern

pufit commented:
> @pufit Ok, this goes in the opposite direction of my last commit. My last change can still be of use for the allow_plaintext_password case.

Yes, it's still useful and solves: https://github.com/ClickHouse/ClickHouse/pull/65277#discussion_r1667053180. Adding some warnings to the logs should also be good, though.

> There is one use case for allowing NO_PASSWORD to co-exist: a grace period for a user that now has password.
It also simplifies things. After all, we don't need any sort of drop upon authentication method registration.

The major problem is that your example [here](https://github.com/ClickHouse/ClickHouse/pull/65277#issuecomment-2214755626) will be a valid authentication, which is very counterintuitive and can lead to potential security issues.

arthurpassos commented:
@pufit Can you elaborate on the below?

> We should not allow adding the NO_PASSWORD method to existing ones, so an error should probably be enough. If a user wants to set NO_PASSWORD, they should override auth methods explicitly.

From what I can tell, you don't want to allow it. But then you sai the user should override it explicitly, how would the user do that?

arthurpassos commented:
By the way, is there anything else you want me to change before merging?

arthurpassos commented:
There is also the below case:

`ALTER USER user1 ADD IDENTIFIED WITH plaintext_password by '1' ADD IDENTIFIED WITH no_password ADD IDENTIFIED_WITH bcrypt_password '2'`. I suppose it should not be allowed.

arthurpassos committed:
do not allow no_password to co-exist with other auth methods

arthurpassos commented:
Implemented your suggestion, let me know what you think @pufit

pufit commented:
> > @pufit Can you elaborate on the below?
>
> We should not allow adding the NO_PASSWORD method to existing ones, so an error should probably be enough. If a user wants to set NO_PASSWORD, they should override auth methods explicitly.
>
>From what I can tell, you don't want to allow it. But then you sai the user should override it explicitly, how would the user do that?

I mean if you want to set a user auth method NO_PASSWORD, you should run
`ALTER USER user1 IDENTIFIED WITH no_password`

And
`ALTER USER user1 ADD IDENTIFIED WITH no_password`
should result in an error if any other auth methods were set.

> By the way, is there anything else you want me to change before merging?

As far as I can tell everything looks good

arthurpassos committed:
do not allow no_pwd to co-exist with other auth methods

arthurpassos commented:
> > > @pufit Can you elaborate on the below?
> > 
> > 
> > We should not allow adding the NO_PASSWORD method to existing ones, so an error should probably be enough. If a user wants to set NO_PASSWORD, they should override auth methods explicitly.
> > From what I can tell, you don't want to allow it. But then you sai the user should override it explicitly, how would the user do that?
> 
> I mean if you want to set a user auth method NO_PASSWORD, you should run `ALTER USER user1 IDENTIFIED WITH no_password`
> 
> And `ALTER USER user1 ADD IDENTIFIED WITH no_password` should result in an error if any other auth methods were set.
> 
> > By the way, is there anything else you want me to change before merging?
> 
> As far as I can tell everything looks good

Implemented it :+1:

filimonov commented on docs/en/sql-reference/statements/create/user.md:
````diff
@@ -144,6 +144,12 @@ In ClickHouse Cloud, by default, passwords must meet the following complexity re
+7. Multiple authentication methods can be specified: 
+
+   ```sql
+   CREATE USER user1 IDENTIFIED WITH plaintext_password by '1' ADD IDENTIFIED WITH bcrypt_password by '2' ADD IDENTIFIED WITH plaintext_password by '3''
````
Syntax with ADD looks odd here.

Using verb 'add' in the middle of CREATE USER ... is strange. Maybe just comma separated?

filimonov replied:
ALTER ... ADD looks ok. 

But CREATE ... ADD - is weird.

arthurpassos replied:
@pufit I have discussed in private with @filimonov about the syntax and we thought it would be better to simplify it to a comma-separated list like the below:

`CREATE USER user1 IDENTIFIED WITH plaintext_password by '1', sha256_password by '2', bcypt_password by '3'...`
`ALTER USER user1 IDENTIFIED WITH plaintext_password by '1', sha256_password by '2', bcypt_password by '3'...`
`ALTER USER user1 ADD IDENTIFIED WITH plaintext_password by '1', sha256_password by '2', bcypt_password by '3'...`

As you probably know, the `WITH` keyword can be omitted and the authentication type does not need to be specified. In this case, I opted for the following:

`CREATE USER user1 IDENTIFIED by '1', plaintext_password by '2', by '3'` -- If the default auth type is sha256, then the user has two sha256 auth methods and one plaintext.

@alexey-milovidov fyi

filimonov commented on src/Access/Authentication.cpp:
```diff
@@ -84,218 +79,158 @@ namespace
+        switch (authentication_method.getType())
         {
             case AuthenticationType::NO_PASSWORD:
-                return true; // N.B. even if the password is not empty!
```
return the comment back, seems important

filimonov commented on src/Access/IAccessStorage.cpp:
```diff
@@ -747,14 +757,6 @@ void IAccessStorage::throwAddressNotAllowed(const Poco::Net::IPAddress & address
-        ErrorCodes::AUTHENTICATION_FAILED,
-        "Authentication type {} is not allowed, check the setting allow_{} in the server configuration",
-        toString(auth_type), AuthenticationTypeInfo::get(auth_type).name);
-}
```
May be that should still go the logs? 

BTW showing that exception to the user (i.e. how it is right now) disclosures some information about the user (i.e. the fact that it exists & have bad auth schema configured). 

But that information can be important for admins / DBAs.

arthurpassos replied:
Added in the form of `LOG_INFO`

arthurpassos commented on src/Access/IAccessStorage.cpp:
```diff
@@ -525,15 +524,26 @@ std::optional<AuthResult> IAccessStorage::authenticateImpl(
+            // for now, just throw exception in case a user exists with invalid auth method
+            // back in the day, it would also throw an exception. There might be a smarter alternative
+            // like a user scan during startup.
+            for (const auto & auth_method : user->authentication_methods)
```
It should. IIRC, serialization and de-serialization happens in order.

arthurpassos committed:
use comma separated instead of multiple add id add id

arthurpassos committed:
rename test and add a few more

arthurpassos committed:
add comment back and missing file

arthurpassos committed:
fix hilite test

arthurpassos committed:
trigger ci

arthurpassos committed:
Removed stale comment and added log_info about skipped auth methods

arthurpassos committed:
fix a few ut

arthurpassos committed:
update docs

arthurpassos commented:
@pufit ping for thoughts on https://github.com/ClickHouse/ClickHouse/pull/65277#discussion_r1674648266

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
remove dupl inc

alexey-milovidov commented:
We have noticed a failure of this test in master: https://s3.amazonaws.com/clickhouse-test-reports/0/30725bcd7a9eab3332ef15b2f6d84e9226d0d38e/integration_tests__tsan__[2_6].html

arthurpassos commented:
> We have noticed a failure of this test in master: [https://s3.amazonaws.com/clickhouse-test-reports/0/30725bcd7a9eab3332ef15b2f6d84e9226d0d38e/integration_tests__tsan__[2_6].html](https://s3.amazonaws.com/clickhouse-test-reports/0/30725bcd7a9eab3332ef15b2f6d84e9226d0d38e/integration_tests__tsan__%5B2_6%5D.html)

Are you pinging me because you want my help to investigate this test or because you think I was the author of it? I am willing to help, but just to be clear, I did not create this test.

alexey-milovidov commented:
Sorry for pinging. The test has another author, @ianton-ru

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
fix conflict

arthurpassos committed:
fix test that was recently introduced

pufit commented:
Hi, @arthurpassos! Sorry for the delay in the review. I took a look at the implemented changes, and everything looks good to me! Can you just add some more comments about this behavior https://github.com/ClickHouse/ClickHouse/pull/65277#issuecomment-2219579070 to the docs?

arthurpassos committed:
add notes about no_password behavior

arthurpassos committed:
remove todo

arthurpassos commented on src/Storages/System/StorageSystemUsers.cpp:
```diff
@@ -119,6 +119,7 @@ void StorageSystemUsers::fillData(MutableColumns & res_columns, ContextPtr conte
     auto & column_grantees_except_offsets = assert_cast<ColumnArray &>(*res_columns[column_index++]).getOffsets();
     auto & column_default_database = assert_cast<ColumnString &>(*res_columns[column_index++]);
 
+    // todo arthur check this
```
I forgot to look into this one, gimme a few

arthurpassos replied:
Hm.. this one might be a bit trickier

arthurpassos replied:
@pufit do you have an opinion on how this should evolve? If I understand correctly, this is about the users table. Should we turn the `auth_type` column into a list and add more json objects on `auth_data`?

arthurpassos replied:
I converted the `auth_type` column from `Int8` to `Array(Int8)` and kept `auth_params` as a `String`, but populated it with a JSON Array instead of a single JSON object.

It might make more sense a be more idiomatic to make it an array as well..

arthurpassos replied:
Changed it to an array of strings=

pufit replied:
> I forgot to look into this one, gimme a few

Yes, I missed it during the review as well (kinda awkward 😅)

> It might make more sense a be more idiomatic to make it an array as well..

Yes, I think `Array(Int8)` and `Array(String)` should work

pufit replied:
Can you also add a test on system.users with multiple auth methods?

arthurpassos replied:
Fixed existing settings and added a new one

arthurpassos committed:
make auth_type a vector of int8_t and auth_params a json array

arthurpassos committed:
change auth_params type from string to array<string>

arthurpassos commented:
@alexey-milovidov @pufit what is your take on breaking downgrades? I haven't tested it myself, but if a user was created with multiple authentication methods and then the server is downgraded to a version that supports it, I believe it won't be able to boot.

Btw, this should also happen with other features previously implemented like https://github.com/ClickHouse/ClickHouse/pull/51261/

arthurpassos committed:
update test

alexey-milovidov commented:
PTAL https://github.com/ClickHouse/ClickHouse/pull/53286#issuecomment-2251416434

arthurpassos committed:
fix tests and add new one

arthurpassos committed:
style

arthurpassos commented:
@pufit In case a user with multiple authentication methods was created and the server is downgraded, that user will become unusable, along with some other features like `show users`. This happens because the query serialized now contains new syntax, which is considererd invalid by previous versions. See example below:

ClickHouse in a version that supports this feature, there is a user with multiple authentication methods.
```
arthur :) select name, id, auth_type, auth_params from system.users;

SELECT
    name,
    id,
    auth_type,
    auth_params
FROM system.users

Query id: 69638b9e-c18c-48c9-a5e6-acae57ddd184

Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 24.7.1.

   ┌─name──────────────────────┬─id───────────────────────────────────┬─auth_type──────────────────────────────────────────────────┬─auth_params──────┐
1. │ user1                     │ 0b429503-1769-6cee-3d98-302ded3b388d │ ['plaintext_password','sha256_password','sha256_password'] │ ['{}','{}','{}'] │
2. │ default                   │ 94309d50-4f52-5250-31bd-74fecac179db │ ['plaintext_password']                                     │ ['{}']           │
3. │ session_log_test_xml_user │ fbfbcb2a-52b7-6d93-3ac0-95ca947685ab │ ['no_password']                                            │ ['{}']           │
4. │ readonly                  │ 97db8f60-4e69-f266-a0d9-c5270500251e │ ['plaintext_password']                                     │ ['{}']           │
   └───────────────────────────┴──────────────────────────────────────┴────────────────────────────────────────────────────────────┴──────────────────┘
```

After downgrading:

```
arthur :) select name, id, auth_type, auth_params from system.users;

SELECT
    name,
    id,
    auth_type,
    auth_params
FROM system.users

Query id: ccb3e37d-6cf8-4632-b468-3d50fc7979bb

Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 24.7.1.


Elapsed: 0.044 sec. 

Received exception from server (version 24.7.1):
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Syntax error: failed at position 60 (','): , sha256_hash BY 'D4735E3A265E16EEE03F59718B9B5D03019C07D8B6C51F90DA3A666EEC13AB35', sha256_hash BY '2D6B723522FD48EB6F4B7763BECF161C4339B9E6ED71C430D687886A5B6. Expected one of: VALID UNTIL, HOST, SETTINGS, DEFAULT ROLE, ON, GRANTEES, DEFAULT DATABASE, IN, end of query: Could not parse /home/arthur/ClickHouse/programs/server/access/0b429503-1769-6cee-3d98-302ded3b388d.sql. (SYNTAX_ERROR)
```

Can't connect with that user either:

```
arthur@arthur:~/ClickHouse$ cmake-build-debug/programs/clickhouse client -mn --user user1
ClickHouse client version 24.7.1.1.
Connecting to localhost:9000 as user user1.
Code: 62. DB::Exception: Received from localhost:9000. DB::Exception: Syntax error: failed at position 60 (','): , sha256_hash BY 'D4735E3A265E16EEE03F59718B9B5D03019C07D8B6C51F90DA3A666EEC13AB35', sha256_hash BY '2D6B723522FD48EB6F4B7763BECF161C4339B9E6ED71C430D687886A5B6. Expected one of: VALID UNTIL, HOST, SETTINGS, DEFAULT ROLE, ON, GRANTEES, DEFAULT DATABASE, IN, end of query: Could not parse /home/arthur/ClickHouse/programs/server/access/0b429503-1769-6cee-3d98-302ded3b388d.sql. (SYNTAX_ERROR)
```

Dropping the user does the trick:

```
arthur :) drop user user1;

DROP USER user1

Query id: 668181d6-4dbc-4763-b9f3-9ade2d53dcb2

Ok.

0 rows in set. Elapsed: 0.002 sec. 
```

```
arthur :) select name, id, auth_type, auth_params from system.users;

SELECT
    name,
    id,
    auth_type,
    auth_params
FROM system.users

Query id: a9c5e7d2-1212-491c-a66f-440eb7526c26

   ┌─name──────────────────────┬─id───────────────────────────────────┬─auth_type──────────┬─auth_params─┐
1. │ default                   │ 94309d50-4f52-5250-31bd-74fecac179db │ plaintext_password │ {}          │
2. │ session_log_test_xml_user │ fbfbcb2a-52b7-6d93-3ac0-95ca947685ab │ no_password        │ {}          │
3. │ readonly                  │ 97db8f60-4e69-f266-a0d9-c5270500251e │ plaintext_password │ {}          │
   └───────────────────────────┴──────────────────────────────────────┴────────────────────┴─────────────┘

3 rows in set. Elapsed: 0.003 sec. 
```

Note: the downgrade itself will not cause this issue. This will only happen if there is a downgrade and a user with multiple authentication methods. Users with a single authentication method will keep working just fine.

In order to gracefully downgrade, all users must have only one authentication methods. Where should we document this procedure?

arthurpassos committed:
add note about downgrading

arthurpassos committed:
minor doc change

arthurpassos committed:
small comment update

arthurpassos committed:
update logout session

arthurpassos committed:
update some integ tests

arthurpassos committed:
retrigger ci

arthurpassos commented:
@pufit Did I break docs somehow? Docs check seems to be failing, but I don't see any meaningful output on https://s3.amazonaws.com/clickhouse-test-reports/65277/c433d9cfdb321f27289df4f35f434a2685af890d/docs_check/run.log

arthurpassos committed:
try to fix docs? is it broken at all

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos commented:
> @pufit Did I break docs somehow? Docs check seems to be failing, but I don't see any meaningful output on https://s3.amazonaws.com/clickhouse-test-reports/65277/c433d9cfdb321f27289df4f35f434a2685af890d/docs_check/run.log

Fixed

arthurpassos commented:
@pufit Can you assist me on fixing the CI run? I see two failures: flaky abd stress. Latter is already reported: https://github.com/ClickHouse/ClickHouse/issues/67330.

The former, tho, I can't figure out what is wrong and why is it being considered flaky. I have analyzed several log files: `flaky_0.log`, `flaky_0.nfo`, `flaky_0_pytest-gw0.log` and others. None of them reveal the cause for the failure. Maybe the failure is not related to my changes at all, but to the fact that this test was never executed against this version of flaky test and it is now being executed because I made some changes to it?

Regardless, how do we identify the cause / proceed?

arthurpassos commented:
> @pufit Can you assist me on fixing the CI run? I see two failures: flaky abd stress. Latter is already reported: #67330.
> 
> The former, tho, I can't figure out what is wrong and why is it being considered flaky. I have analyzed several log files: `flaky_0.log`, `flaky_0.nfo`, `flaky_0_pytest-gw0.log` and others. None of them reveal the cause for the failure. Maybe the failure is not related to my changes at all, but to the fact that this test was never executed against this version of flaky test and it is now being executed because I made some changes to it?
> 
> Regardless, how do we identify the cause / proceed?

Found something interesting:

```
2024.07.31 13:21:10.878603 [ 10 ] {aecfbd03-12f5-4ffc-a828-ef0aa84fd496} <Fatal> : Logical error: 'Inconsistent AST formatting: the query:
GRANT CURRENT GRANTS TO C
cannot parse.'.
2024.07.31 13:21:10.933067 [ 10 ] {aecfbd03-12f5-4ffc-a828-ef0aa84fd496} <Fatal> : Stack trace (when copying this message, always include the lines below):

0. /home/arthur/ClickHouse/contrib/llvm-project/libcxx/include/exception:141: std::exception::capture() @ 0x000000000ae247a2
1. /home/arthur/ClickHouse/contrib/llvm-project/libcxx/include/exception:116: std::exception::exception[abi:v15007]() @ 0x000000000ae2476d
2. /home/arthur/ClickHouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x0000000023317d80
3. /home/arthur/ClickHouse/src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013a7b951
4. /home/arthur/ClickHouse/src/Common/Exception.h:110: DB::Exception::Exception(String&&, int, bool) @ 0x000000000ae1a7ea
5. /home/arthur/ClickHouse/src/Common/Exception.h:65: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000ae17e29
6. /home/arthur/ClickHouse/src/Common/Exception.h:128: DB::Exception::Exception<String&>(int, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x000000000ae91165
7. /home/arthur/ClickHouse/src/Interpreters/executeQuery.cpp:800: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001bdc9ebe
8. /home/arthur/ClickHouse/src/Interpreters/executeQuery.cpp:1390: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000001bdc87aa
9. /home/arthur/ClickHouse/src/Server/TCPHandler.cpp:523: DB::TCPHandler::runImpl() @ 0x000000001dd34027
10. /home/arthur/ClickHouse/src/Server/TCPHandler.cpp:2389: DB::TCPHandler::run() @ 0x000000001dd499df
11. /home/arthur/ClickHouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x00000000233fc5f9
12. /home/arthur/ClickHouse/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x00000000233fce48
13. /home/arthur/ClickHouse/base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x000000002338b5ce
14. /home/arthur/ClickHouse/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x0000000023387cda
15. /home/arthur/ClickHouse/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000023386a1e
16. ? @ 0x0000784aafc27ac3
17. ? @ 0x0000784aafcb9850
```

arthurpassos commented:
> > @pufit Can you assist me on fixing the CI run? I see two failures: flaky abd stress. Latter is already reported: #67330.
> > The former, tho, I can't figure out what is wrong and why is it being considered flaky. I have analyzed several log files: `flaky_0.log`, `flaky_0.nfo`, `flaky_0_pytest-gw0.log` and others. None of them reveal the cause for the failure. Maybe the failure is not related to my changes at all, but to the fact that this test was never executed against this version of flaky test and it is now being executed because I made some changes to it?
> > Regardless, how do we identify the cause / proceed?
> 
> Found something interesting:
> 
> ```
> 2024.07.31 13:21:10.878603 [ 10 ] {aecfbd03-12f5-4ffc-a828-ef0aa84fd496} <Fatal> : Logical error: 'Inconsistent AST formatting: the query:
> GRANT CURRENT GRANTS TO C
> cannot parse.'.
> 2024.07.31 13:21:10.933067 [ 10 ] {aecfbd03-12f5-4ffc-a828-ef0aa84fd496} <Fatal> : Stack trace (when copying this message, always include the lines below):
> 
> 0. /home/arthur/ClickHouse/contrib/llvm-project/libcxx/include/exception:141: std::exception::capture() @ 0x000000000ae247a2
> 1. /home/arthur/ClickHouse/contrib/llvm-project/libcxx/include/exception:116: std::exception::exception[abi:v15007]() @ 0x000000000ae2476d
> 2. /home/arthur/ClickHouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x0000000023317d80
> 3. /home/arthur/ClickHouse/src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013a7b951
> 4. /home/arthur/ClickHouse/src/Common/Exception.h:110: DB::Exception::Exception(String&&, int, bool) @ 0x000000000ae1a7ea
> 5. /home/arthur/ClickHouse/src/Common/Exception.h:65: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000ae17e29
> 6. /home/arthur/ClickHouse/src/Common/Exception.h:128: DB::Exception::Exception<String&>(int, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x000000000ae91165
> 7. /home/arthur/ClickHouse/src/Interpreters/executeQuery.cpp:800: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001bdc9ebe
> 8. /home/arthur/ClickHouse/src/Interpreters/executeQuery.cpp:1390: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x000000001bdc87aa
> 9. /home/arthur/ClickHouse/src/Server/TCPHandler.cpp:523: DB::TCPHandler::runImpl() @ 0x000000001dd34027
> 10. /home/arthur/ClickHouse/src/Server/TCPHandler.cpp:2389: DB::TCPHandler::run() @ 0x000000001dd499df
> 11. /home/arthur/ClickHouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x00000000233fc5f9
> 12. /home/arthur/ClickHouse/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x00000000233fce48
> 13. /home/arthur/ClickHouse/base/poco/Foundation/src/ThreadPool.cpp:205: Poco::PooledThread::run() @ 0x000000002338b5ce
> 14. /home/arthur/ClickHouse/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x0000000023387cda
> 15. /home/arthur/ClickHouse/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000023386a1e
> 16. ? @ 0x0000784aafc27ac3
> 17. ? @ 0x0000784aafcb9850
> ```

Not related to my PR, more details in: https://github.com/ClickHouse/ClickHouse/commit/078b8f53990072ad797488a1f66f57390f0cf14c#diff-9d106b2999421e27ea1924830817d4b8a6ea9fc093f797fd30f9a592fe9f338eR616

arthurpassos commented:
@pufit I honestly couldn't find why this is failing. Waiting for your input

pufit commented:
> Note: the downgrade itself will not cause this issue. This will only happen if there is a downgrade and a user with multiple authentication methods. Users with a single authentication method will keep working just fine.

This is a common situation when a new syntax is added. We can't do much about that. 

> @pufit I honestly couldn't find why this is failing. Waiting for your input

https://github.com/ClickHouse/ClickHouse/pull/67653

arthurpassos commented:
> > Note: the downgrade itself will not cause this issue. This will only happen if there is a downgrade and a user with multiple authentication methods. Users with a single authentication method will keep working just fine.
> 
> This is a common situation when a new syntax is added. We can't do much about that.
> 
> > @pufit I honestly couldn't find why this is failing. Waiting for your input
> 
> #67653

So.. what is our plan? Wait for that PR to be merged and then merge master? I could not determine if CI/CD errors are related to that one, after all, that test itself does not fail on CI/CD.

pufit commented:
> Wait for that PR to be merged and then merge master?

Yes, let's wait and then rebase once more.

arthurpassos commented:
> > Wait for that PR to be merged and then merge master?
> 
> Yes, let's wait and then rebase once more.

any news on that one?

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
add docs about reset auth method

arthurpassos committed:
fix trailing comma issue

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

alsugiliazova commented:
Is it reasonable to allow unlimited number of authentication methods for a user?
It might be not practical for several reasons:
1. It can be difficult to keep track of all the methods assigned to a user.
2. Security. Each authentication method is a potential entry point. The more methods available, the higher the risk that one of them might be compromised.
3. Most users don't need more than a few authentication methods (e.g., password, certificate, or Kerberos). It is hard to test unlimited amount of combinations.

filimonov commented:
> Is it reasonable to allow unlimited number of authentication methods for a user? It might be not practical for several reasons:

We can introduce server-level setting limiting that. Smth like `max_authentication_methods_per_user`

arthurpassos commented:
> Is it reasonable to allow unlimited number of authentication methods for a user? It might be not practical for several reasons:
> 
> 1. It can be difficult to keep track of all the methods assigned to a user.
> 2. Security. Each authentication method is a potential entry point. The more methods available, the higher the risk that one of them might be compromised.
> 3. Most users don't need more than a few authentication methods (e.g., password, certificate, or Kerberos). It is hard to test unlimited amount of combinations.

@pufit @alexey-milovidov thoughts?

pufit commented:
> @pufit @alexey-milovidov thoughts?

We can assume that if a user has rights to access management (which includes `ALTER USER`), this user should know what they are doing and understand the risk of adding multiple auth methods. Nevertheless, I think adding the `max_authentication_methods_per_user` server settings is a good improvement. This setting will allow the user to turn off the multi-auth feature if necessary. But we can set it to unlimited by default.

arthurpassos commented:
> > @pufit @alexey-milovidov thoughts?
> 
> We can assume that if a user has rights to access management (which includes `ALTER USER`), this user should know what they are doing and understand the risk of adding multiple auth methods. Nevertheless, I think adding the `max_authentication_methods_per_user` server settings is a good improvement. This setting will allow the user to turn off the multi-auth feature if necessary. But we can set it to unlimited by default.

I am ok with having the settings as well. I am more inclined on leaving unlimited on default or set to a high number. If a malicious user has rights to create authentication methods, it is already very bad. It also sounds a bit weird to me "we support multiple authentication methods per user, but by default, only three, which doesn't even cover all authentication types available".

In any case, that's just my opinion, I am fine with either approaches. The end result will be very similar, I think.

arthurpassos commented:
There is also the question on how we would handle changes on this setting. Consider there is a user `A` with 10 authentication methods. Now the setting is changed to allow only 5 authentication methods. What is expected to happen? 

1. Succeed to set the setting, but whenever user `A` tries to connect, throw an exception saying it should be fixed.
2. Fail to set the setting? Not sure it is possible to implement or if there is a similar case in ClickHouse, but probably yes.

pufit commented:
> There is also the question on how we would handle changes on this setting. Consider there is a user `A` with 10 authentication methods. Now the setting is changed to allow only 5 authentication methods. What is expected to happen?
> 
> 1. Succeed to set the setting, but whenever user `A` tries to connect, throw an exception saying it should be fixed.
> 2. Fail to set the setting? Not sure it is possible to implement or if there is a similar case in ClickHouse, but probably yes.

I think the best approach is to ignore this. During `ALTER/CREATE USER,` we can check if the total expected number of the auth methods is larger than set. But if, for some reason, we create a user with multiple auth methods and then change the setting to 1, the already created users shouldn't be affected by this constraint.

alsugiliazova commented:
Adding a max_authentication_methods_per_user setting is a good idea. There are three main options for the default value:

1. Default to 2 or 3: This provides a reasonable baseline, since multi-authentication is primarily for scenarios like password rotation or dual password support.

2. Default to 1: This matches the current behavior, with users able to increase the limit to whatever number they need, up to a **specified maximum**, which is also should be discussed. This option allows users to set the limit according to their specific use case.

3. Default to some big number ("unlimited"). 

Let’s finalize the PR with one of these options.  Any further feedback is welcome.

arthurpassos committed:
fix leading id method without WITH and with type specified being allowed

arthurpassos committed:
make sure reset authentication methods can only be used on alter queries

arthurpassos committed:
add server setting

arthurpassos committed:
add integ tests for new setting

arthurpassos committed:
add integ tests for defa7ult value

arthurpassos committed:
fix black

arthurpassos committed:
fix black with script

arthurpassos commented:
Added a setting with default value to `256`, which is high enough and not "unlimited". Upon changing this setting, existing users do not get affected. It is only checked upon `ALTER` or `CREATE` statements.

@pufit let me know what you think

alexey-milovidov commented:
256 sounds special and typically implies that alignment benefits for something, but it is not the case.
Make it 100 by default.

arthurpassos committed:
change default setting value to 100

arthurpassos commented:
> 256 sounds special and typically implies that alignment benefits for something, but it is not the case. Make it 100 by default.

Done. @pufit could you please have a look?

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos commented:
@pufit Hi. I kindly ask for your help on flaky test failures, it happened again and I simply can't tell what is going on. Those are all old tests that had to be updated due to the new logic introduced in this PR

arthurpassos committed:
fix serialization to fix on cluster commands

arthurpassos committed:
fix in the right place..

arthurpassos committed:
throw syntax error instead of bad arguments in case of add not identified

arthurpassos committed:
exception message

arthurpassos committed:
add on cluster tests

arthurpassos committed:
do not allow reset auth to new to be used along with add identified clauses

arthurpassos commented:
> @pufit Hi. I kindly ask for your help on flaky test failures, it happened again and I simply can't tell what is going on. Those are all old tests that had to be updated due to the new logic introduced in this PR

@pufit ping for assistance

arthurpassos committed:
make sure all alters except id related are allowed even if state of user is invalid afterwards

arthurpassos committed:
make max_number_of_authentication_methods=0 unlimited

arthurpassos committed:
add setting docs

arthurpassos committed:
minor grammar

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
Update 03174_multiple_authentication_methods.sh

arthurpassos committed:
change parsing logic a bit

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
style

arthurpassos committed:
remove extra dot from ex message

arthurpassos committed:
extract some tests into sql as an attempt to make them run faster for flaky check..

arthurpassos committed:
add no fast test

arthurpassos committed:
no parallel

arthurpassos committed:
no replicated database

arthurpassos committed:
add space

arthurpassos commented:
@pufit Flaky check run is still timing out. I have moved the tests that do not depend on `clickhouse-client` and login functionality out of the `.sh` file to a `.sql` file and it helped a bit. The `.sh` runs dropped from 100~110 seconds to 50~65 seconds, while the `.sql` runs are now taking 10~20 seconds.

What do you suggest? I can remove the `on cluster` tests, this would drop the execution time by half. I don't think this is a good idea, tho. Maybe the flaky check timeout is poorly calculated? What if there is a PR that touches several test files, there is a high chance this flaky run will fail for no good reason.

https://s3.amazonaws.com/clickhouse-test-reports/65277/41a4a97ca3c47bc4420b9247572f937569b98e6f/stateless_tests_flaky_check__asan_.html

arthurpassos committed:
use curl client for create/drop/alter queries to make test run faster

arthurpassos committed:
further optimize test by using http client instead of clickhouse-client

arthurpassos committed:
merge test fils now that they are faster

arthurpassos committed:
add extra line

arthurpassos committed:
fix test

arthurpassos committed:
remove non clustered tests to make it faster :D

arthurpassos committed:
add head -n 1 to make sure only on occurrence of error code is grepped

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
split tests again and randomize user name to be able to parallel

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos committed:
add cleanup code to existing tests so flaky tests pass..

arthurpassos committed:
add new line

arthurpassos commented:
@pufit CI appears to be broken for quite some time already https://github.com/ClickHouse/ClickHouse/pull/69080, and several non-CI related PRs are being merged regardless. I have managed to make the tests run faster, can we merge this?

nikitamikhaylov commented:
@arthurpassos We can accept this PR only in case if it is 100% green with all 200+ checks passed.

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos commented:
> @arthurpassos We can accept this PR only in case if it is 100% green with all 200+ checks passed. We should carefully iterate over all tests that failed, try to fix them or create a very informative issue to make other contributors lives easier (there two issues #69127 #69126 are not very helpful, because they don't tell **which** tests failed).

is it green enoug now?

arthurpassos commented:
@alexey-milovidov @nikitamikhaylov ping

nikitamikhaylov commented:
Looks much greener, yes!

arthurpassos commented:
> Looks much greener, yes!

Good, can we merge it now?

arthurpassos commented:
> The CI says it is not ready yet.

Is this about performance check failure? It has been some time since I worked on a repo I had merge permissions, so I can't really tell what it is complaining about

arthurpassos commented:
> **Mergeable Check ** Pending — pending: Cloud fork sync (only for ClickHouse Inc. employees)

@nikitamikhaylov is this what is blocking the merge? If so, what needs to be done? Anything I should do on my side?

arthurpassos commented:
@pufit kind ping

pufit commented:
> @pufit kind ping

The only thing left is to fix conflicts in the cloud repo on our side. I will do it by the end of Friday. Sorry for waiting

arthurpassos committed:
Merge branch 'master' into multi_auth_methods

arthurpassos commented:
> > @pufit kind ping
> 
> The only thing left is to fix conflicts in the cloud repo on our side. I will do it by the end of Friday. Sorry for waiting

https://github.com/ClickHouse/ClickHouse/pull/68236 has been merged and introduced some conflicts on my side. Just fixed them.

arthurpassos committed:
update recently introduced test

arthurpassos committed:
fix black

arthurpassos committed:
trigger ic

arthurpassos commented:
Got CI/CD green again
